DTS: work towards replace generated conf with EDT

galak opened the pull request.

galak committed:
kconfig: nxp_imx: rt: Convert to use new kconfigfunctions

Convert use of dt_int_val/dt_hex_val to dt_node_reg_{addr,size}.  As
we want to remove dt_int_val/dt_hex_val.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
kconfig: counter: nrfx: Convert to use new kconfigfunctions

Convert use of dt_int_val to dt_node_has_bool_prop.  As we want to
remove dt_int_val.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
kconfig: Convert device tree chosen properties to new kconfigfunctions

Convert how we get the various chosen properties like "zephyr,console"
to use the new kconfig functions like dt_chosen_to_label.

Because of how kconfig parses things we define a set of variables of the
form DT_CHOSEN_Z_<PROP> since comma's are parsed as field seperators in
macros.

This conversion allows us to remove code in gen_defines.py for the
following chosen properties:

zephyr,console
zephyr,shell-uart
zephyr,bt-uart
zephyr,uart-pipe
zephyr,bt-mon-uart
zephyr,uart-mcumgr
zephyr,bt-c2h-uart

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
kconfig: Add new functions to kconfigfunctions to use EDT

Add a new set of functions that utilize EDT so we can move away from the
generated .conf file.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
kconfig: hifive1_revb: Convert to use new kconfigfunctions

Convert use of dt_hex_val to dt_node_reg_addr.  As we want to
remove dt_hex_val.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

carlescufi commented on Kconfig.zephyr:
```diff
@@ -94,9 +94,11 @@ config USE_CODE_PARTITION
 	help
 	  When selected application will be linked into chosen code-partition.
 
+DT_CHOSEN_Z_CODE_PARTITION := zephyr,code-partition
```
What sourcery is this? Is this standard Kconfig? And if so, why not use the `zephyr,code-partition` string directly in the call to `dt_chosen_to_reg_addr()`?

carlescufi replied:
I did not read the commit message :)
Thanks for the details

carlescufi commented on boards/riscv/hifive1_revb/Kconfig.defconfig:
```diff
@@ -9,7 +9,7 @@ config HAS_FLASH_LOAD_OFFSET
 
 config FLASH_BASE_ADDRESS
-	default $(dt_hex_val,DT_FLASH_BASE_ADDRESS)
+	default $(dt_path_to_reg_addr,/soc/spi@10014000,1)
```
is this right?

galak committed:
kconfig: serial: sam0: Convert to use new kconfigfunctions

Convert use of dt_int_val to dt_compat_enabled.  As we want to
remove dt_int_val.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

ulfalizer commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -33,66 +34,198 @@ def _dt_units_to_scale(unit):
+    device has a "label" property and return the value of that "label", if not
+    we return an empty string.
+    """
+
```
No blank line after `"""` to be consistent.

ulfalizer commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -33,66 +34,198 @@ def _dt_units_to_scale(unit):
+        return ""
+
+    return dev.props["label"].val
+
```
Missing blank line.

ulfalizer commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -5,6 +5,19 @@
+import edtlib
+
+DTS_POST_CPP = os.environ["DTS_POST_CPP"]
+DTS_ROOT_BINDINGS = os.environ.get("DTS_ROOT_BINDINGS")
```
`BINDINGS_DIR` might be better (was thinking "what's a root binding, and how is it different from a regular binding?")

Might be nice to have a comment that explains what it means when it's unset, and when it would happen.

ulfalizer commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -91,8 +104,172 @@ def dt_str_val(kconf, _, name):
+    try:
+       dev = edt.get_dev(path)
+    except edtlib.EDTError:
+       return "0"
```
Typo'd indent (3 instead of 4).

ulfalizer commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -91,8 +104,172 @@ def dt_str_val(kconf, _, name):
+        if dev.enabled:
+            return "y"
+
+    return "n"
```
Could remove `compat_devs()` and do

```python
for dev in edt.devices:
    if compat in dev.compats and dev.enabled:
        return "y"

return "n"
```

It'll be fast enough.

galak replied:
Done.

ulfalizer commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -91,8 +104,172 @@ def dt_str_val(kconf, _, name):
+
+    return dev.props["label"].val
+
+def _dt_to_reg_addr(dev, index, unit):
```
`_dev_reg_addr`?

Sounded like it mapped device trees to register addresses first.

galak replied:
Changed.

ulfalizer commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -91,8 +104,172 @@ def dt_str_val(kconf, _, name):
+    return hex(dev.regs[int(index)].addr >> _dt_units_to_scale(unit))
+
+
+def _dt_to_reg_size(dev, index, unit):
```
`_dev_reg_size`?

ulfalizer commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -91,8 +104,172 @@ def dt_str_val(kconf, _, name):
+
+    dev = edt.chosen_dev(chosen)
+
+    return _dt_to_reg_addr(dev, index, unit)
```
Could get rid of the single-use variable maybe:

```python
return _dev_reg_addr(edt.chosen_dev(chosen), index, unit)
```

galak replied:
changed.

ulfalizer commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -91,8 +104,172 @@ def dt_str_val(kconf, _, name):
+
+    dev = edt.chosen_dev(chosen)
+
+    return _dt_to_reg_size(dev, index, unit)
```
Ditto here.

ulfalizer commented on drivers/bluetooth/hci/Kconfig:
```diff
@@ -68,9 +68,11 @@ config BT_NO_DRIVER
 
 endchoice
 
+DT_CHOSEN_Z_BT_UART := zephyr,bt-uart
```
A comment would be nice for these. I'd assume the variable is set because it's used in many places, not just to hack around a comma.

Could have a

    comma := ,

definition somewhere instead maybe, and use that.

ulfalizer replied:
Up to you. Could add a short comment like "# Workaround for comma in value" instead.

ulfalizer replied:
Still missing a comment.

I'm about to push a Kconfiglib update that exposes the parsing location as `kconf.filename/linenr` btw, for use in preprocessor functions.

ulfalizer commented on scripts/dts/edtlib.py:
```diff
@@ -141,6 +142,17 @@ def chosen_dev(self, name):
+        if compat not in self._compat2dev:
+            return []
+
+        return self._compat2dev[compat]
```
Can probably skip this function and `_compat2dev` (see comment in `dt_compat_enabled()`).

ulfalizer commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -33,66 +34,198 @@ def _dt_units_to_scale(unit):
+    try:
+        dev = edt.get_dev(path)
+    except edtlib.EDTError:
+       return "n"
```
Bad indent.

ulfalizer commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -33,66 +34,198 @@ def _dt_units_to_scale(unit):
+    """
+    This function takes a 'path' and looks for a EDT device at that path.
+    If it finds a EDT device, it will look to see if that device has a
+    boolean property by the name of 'prop.  If the 'prop' exists it will
```
Missing `'` after `prop`.

ulfalizer commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -33,66 +34,198 @@ def _dt_units_to_scale(unit):
+    return _dt_to_reg_size(dev, index, unit)
+
+
+def dt_path_bool_prop(kconf, _, path, prop):
```
Is it important that it's a boolean property?

Could generalize it a bit and call it `dt_path_has_prop` otherwise.

ulfalizer replied:
Yeah, would only work for stuff declared in bindings now.

Maybe there will be stricter checks against the bindings so it works out to the same thing in practice later at least.

galak replied:
Even with the bindings, since things like interrupts & reg properties get handled via their various classes in EDT we wouldn't be able to have the function test for them if we just used the Property class.

ulfalizer replied:
Alright, fine with me.

Think it could be called `dt_has_bool_prop()`, to get a verb in there?

ulfalizer replied:
Or `dt_path_has_bool_prop()`, if you want to be consistent with the other ones.

Wonder if `node` (or nothing) would've been better than `path`. No worries if it's awkward to change though.

galak replied:
what do you mean by `node`?  Just changing the var name or something else?

ulfalizer replied:
Meant that `dt_node_has_bool_prop` might be better than `dt_path_has_bool_prop`. Not a clear win though...

A `has` would be nice though, so it's not just a noun for a function that tests if the node has a property. Would confuse me.

galak replied:
How do you feel about exposing or using `EDT._dt`?  I would rather this be `dt_path_has_prop`.

ulfalizer replied:
Is it useful to test for non-bool properties as well?

Maybe we could come up with something later if we run into cases where we need to check if other properties exist.

ulfalizer reviewed (changes requested):
Maybe it'd be nice to raise errors in some cases in `kconfigfunctions.py` instead of always failing silently.

Can do `raise Exception(msg)` to raise an error. As an undocumented feature, `kconf._filename` and `kconf._linenr` give the location of the macro being expanded.

Maybe I should make them public and document it. Can use the private ones in the meantime at least.

SebastianBoe commented on cmake/dts.cmake:
```diff
@@ -14,6 +14,7 @@ file(MAKE_DIRECTORY ${PROJECT_BINARY_DIR}/include/generated)
 # See ~/zephyr/doc/dts
 set(GENERATED_DTS_BOARD_UNFIXED_H ${PROJECT_BINARY_DIR}/include/generated/generated_dts_board_unfixed.h)
 set(GENERATED_DTS_BOARD_CONF      ${PROJECT_BINARY_DIR}/include/generated/generated_dts_board.conf)
+set(DTS_POST_CPP      ${PROJECT_BINARY_DIR}/${BOARD}.dts.pre.tmp)
```
alignment

galak replied:
Fixed.

galak committed:
kconfig: Mark dt_{int,hex,str}_val as deprecated

As we work on removing the .conf file that is generated by the dt
scripts, mark dt_{int,hex,str}_val functions as deprecated.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

ulfalizer commented:
@galak 
Once https://github.com/zephyrproject-rtos/zephyr/pull/18913 is in, an error reporting helper that also includes the location could be added to the end of `kconfigfunctions.py` like this:

```python
import kconfiglib
...

def yada(kconf, ...):
    ...
    _err(kconf, "bad stuff")

def err(kconf, msg):
    raise kconfiglib.KconfigError(
        "{}:{}: {}".format(kconf.filename, kconf.linenr, msg))
```

ulfalizer commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -47,6 +62,8 @@ def dt_int_val(kconf, _, name, unit=None):
     if doc_mode or name not in dt_defines:
         return "0"
 
+    print("WARNING: dt_int_val is deprecated.")
```
Might be nice to add a `warn()` helper that sends messages to stderr:

```python
def warn(msg):
    print(msg, file=sys.stderr)
```

ulfalizer replied:
This might be better, so that it gives the location as well (like `err()`):

```python
def warn(kconf, msg):
    print("{}:{}: warning: {}".format(kconf.filename, kconf.linenr, msg))
```

galak replied:
How about we do the warn cleanup post merge of your kconfig change.

ulfalizer replied:
Could maybe do just the first `warn()` version for now.

ulfalizer commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -70,6 +87,8 @@ def dt_hex_val(kconf, _, name, unit=None):
     if doc_mode or name not in dt_defines:
         return "0x0"
 
+    print("WARNING: dt_hex_val is deprecated.")
```
Could use the `warn()` helper here too.

ulfalizer commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -89,10 +108,198 @@ def dt_str_val(kconf, _, name):
     if doc_mode or name not in dt_defines:
         return ""
 
+    print("WARNING: dt_str_val is deprecated.")
```
Could use the `warn()` helper here too.

ulfalizer commented on doc/guides/kconfig/index.rst:
```diff
@@ -837,7 +837,7 @@ Device Tree Related Functions
            'g' or 'G'  divide by 1,073,741,824 (1 << 30)
 
-  dt_str_val(kconf, _, name):
+  dt_chosen_label(kconf, _, name):
```
Would be nice to have documentation for the other ones.

Maybe this page could just refer to `scripts/kconfig/kconfigfunctions.py` for some or all of the functions, to avoid having to duplicate the documentation from the docstrings.

galak replied:
Yeah, the doc commit I wanted to wait on til after all the code was good.  Will see if there's some way of just referencing the docstrings.

ulfalizer commented on cmake/kconfig.cmake:
```diff
@@ -39,6 +39,8 @@ set(ENV{SOC_DIR}   ${SOC_DIR})
 set(ENV{ARCH_DIR}   ${ARCH_DIR})
 set(ENV{GENERATED_DTS_BOARD_CONF} ${GENERATED_DTS_BOARD_CONF})
+set(ENV{DTS_POST_CPP} ${DTS_POST_CPP})
+set(ENV{DTS_ROOT_BINDINGS} ${DTS_ROOT_BINDINGS})
```
Maybe these variables could this be called `BINDINGS_DIR` as well, for consistency.

galak replied:
The issue is we'd end up with something like:

`set(ENV{BINDINGS_DIR} ${DTS_ROOT_BINDINGS})`

ulfalizer replied:
Is `DTS_ROOT_BINDINGS` used elsewhere? Was thinking it could be renamed too.

ulfalizer commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -89,10 +108,198 @@ def dt_str_val(kconf, _, name):
+    return dev.props["label"].val
+
+
+def _dev_reg_addr(dev, index, unit):
```
Missing documentation.

ulfalizer replied:
Ah, nm, it's internal. Could add a doc comment though, but up to you.

ulfalizer reviewed (approved):
Seems fine to me.

Can you think of any stuff in `kconfigfunctions.py` that should generate errors? Bit worried that there might be some silent failures now.

galak commented:
> Seems fine to me.
> 
> Can you think of any stuff in `kconfigfunctions.py` that should generate errors? Bit worried that there might be some silent failures now.

I thought about this, but I don't think so.  The only error conditions would be with getting the EDT loaded/setup.
